### PR TITLE
Redis List C API

### PR DESCRIPTION
**Pull request by ZiyanShi**
### What problem does this PR solve?

Implement List C API for integration with Redis

What's Changed:

Implemented List C API and ListIterator C API
remove redundant List C++ API

Tests 

- Unit test

Side effects

Not measured

### CODE REVIEW

**Comment by JiayuZzz** [comment]
Please resolve build check so we can merge

**Review by JiayuZzz — APPROVED** [review]
LGTM

File: tests/tests.cpp [hunk 1]
```diff
@@ -1384,7 +1381,6 @@ TEST_F(EngineBasicTest, TestList) {
       ASSERT_EQ(sz, list_copy.size());
     }
   };
-
   auto RPush = [&](size_t tid) {
     auto const& key = key_vec[tid];
     auto const& elems = elems_vec[tid];
```
Comment (sherry-1001) [inline tests/tests.cpp:1386]: not used. remove it.

File: tests/tests.cpp [hunk 2]
```diff
@@ -1400,8 +1396,10 @@ TEST_F(EngineBasicTest, TestList) {
 
   auto LPop = [&](size_t tid) {
     auto const& key = key_vec[tid];
+    auto const& elems = elems_vec[tid];
     auto& list_copy = list_copy_vec[tid];
     std::string value_got;
+    size_t sz;
```
Comment (sherry-1001) [inline tests/tests.cpp:1402]: remove it.

File: tests/tests.cpp [hunk 3]
```diff
@@ -1400,8 +1396,10 @@ TEST_F(EngineBasicTest, TestList) {
 
   auto LPop = [&](size_t tid) {
     auto const& key = key_vec[tid];
+    auto const& elems = elems_vec[tid];
```
Comment (sherry-1001) [inline tests/tests.cpp:1399]: not used. please remove it.

File: engine/c.cpp [hunk 4]
```diff
@@ -50,8 +59,8 @@ struct KVDKSortedCollectionConfigs {
 };
 
 static char* CopyStringToChar(const std::string& str) {
-  char* result = reinterpret_cast<char*>(malloc(sizeof(char) * str.size()));
-  memcpy(result, str.data(), sizeof(char) * str.size());
+  char* result = static_cast<char*>(malloc(str.size()));
```
Comment (sherry-1001) [inline engine/c.cpp:62]: why remove sizeof(char*)

File: tests/tests.cpp (hunk 2 above)
Comment (ZiyanShi) [reply tests/tests.cpp:1401]: done

File: tests/tests.cpp (hunk 3 above)
Comment (ZiyanShi) [reply tests/tests.cpp]: done

File: engine/c.cpp (hunk 4 above)
Comment (ZiyanShi) [reply engine/c.cpp:62]: sizeof(char) is always 1 according to C99 standard.

**Pull request merged**